File: packages/runar-go/sdk_provider.go
package runar

import (
	"fmt"
	"strings"

	"github.com/bsv-blockchain/go-sdk/transaction"
)

// ---------------------------------------------------------------------------
// Provider interface
// ---------------------------------------------------------------------------

// Provider abstracts blockchain access for UTXO lookup and broadcast.
type Provider interface {
	// GetTransaction fetches a transaction by its txid.
	GetTransaction(txid string) (*TransactionData, error)

	// Broadcast sends a transaction to the network. Returns the txid.
	Broadcast(tx *transaction.Transaction) (string, error)

	// GetUtxos returns all UTXOs for a given address.
	GetUtxos(address string) ([]UTXO, error)

	// GetContractUtxo finds a UTXO by its script hash (for stateful contract lookup).
	// Returns nil if no UTXO is found with the given script hash.
	GetContractUtxo(scriptHash string) (*UTXO, error)

	// GetNetwork returns the network this provider is connected to.
	GetNetwork() string

	// GetFeeRate returns the current fee rate in satoshis per KB (1000 bytes).
	// BSV standard is 100 sat/KB (0.1 sat/byte).
	GetFeeRate() (int64, error)

	// GetRawTransaction fetches the raw transaction hex by its txid.
	GetRawTransaction(txid string) (string, error)
}

// ---------------------------------------------------------------------------
// MockProvider — in-memory provider for testing
// ---------------------------------------------------------------------------

// MockProvider is an in-memory provider for unit tests and local development.
// It stores transactions and UTXOs that can be injected via helper methods,
// and records all broadcasts for assertion in tests.
type MockProvider struct {
	transactions    map[string]*TransactionData
	rawTransactions map[string]string
	utxos           map[string][]UTXO
	contractUtxos   map[string]*UTXO
	broadcastedTxs  []string
	network         string
	broadcastCount  int
	feeRate         int64
}

// NewMockProvider creates a new MockProvider for the given network.
func NewMockProvider(network string) *MockProvider {
	if network == "" {
		network = "testnet"
	}
	return &MockProvider{
		transactions:    make(map[string]*TransactionData),
		rawTransactions: make(map[string]string),
		utxos:           make(map[string][]UTXO),
		contractUtxos:   make(map[string]*UTXO),
		network:         network,
		feeRate:         100,
	}
}

// AddTransaction injects a transaction into the mock store.
func (m *MockProvider) AddTransaction(tx *TransactionData) {
	m.transactions[tx.Txid] = tx
	if tx.Raw != "" {
		m.rawTransactions[tx.Txid] = tx.Raw
	}
}

// AddUtxo injects a UTXO for the given address.
func (m *MockProvider) AddUtxo(address string, utxo UTXO) {
	m.utxos[address] = append(m.utxos[address], utxo)
}

// AddContractUtxo injects a contract UTXO for lookup by script hash.
func (m *MockProvider) AddContractUtxo(scriptHash string, utxo *UTXO) {
	m.contractUtxos[scriptHash] = utxo
}

// GetBroadcastedTxs returns all raw tx hexes that were broadcast.
func (m *MockProvider) GetBroadcastedTxs() []string {
	return m.broadcastedTxs
}

// GetTransaction fetches a transaction from the mock store.
func (m *MockProvider) GetTransaction(txid string) (*TransactionData, error) {
	tx, ok := m.transactions[txid]
	if !ok {
		return nil, fmt.Errorf("MockProvider: transaction %s not found", txid)
	}
	return tx, nil
}

// Broadcast records the transaction and returns a deterministic fake txid.
func (m *MockProvider) Broadcast(tx *transaction.Transaction) (string, error) {
	rawTx := tx.Hex()
	m.broadcastedTxs = append(m.broadcastedTxs, rawTx)
	m.broadcastCount++
	prefix := rawTx
	if len(prefix) > 16 {
		prefix = prefix[:16]
	}
	fakeTxid := mockHash64(fmt.Sprintf("mock-broadcast-%d-%s", m.broadcastCount, prefix))

	// Auto-store raw hex for subsequent getRawTransaction lookups
	if _, ok := m.transactions[fakeTxid]; !ok {
		m.rawTransactions[fakeTxid] = rawTx
	}

	return fakeTxid, nil
}

// GetUtxos returns UTXOs for the given address from the mock store.
func (m *MockProvider) GetUtxos(address string) ([]UTXO, error) {
	return m.utxos[address], nil
}

// GetContractUtxo returns a UTXO by script hash from the mock store.
func (m *MockProvider) GetContractUtxo(scriptHash string) (*UTXO, error) {
	utxo, ok := m.contractUtxos[scriptHash]
	if !ok {
		return nil, nil
	}
	return utxo, nil
}

// GetNetwork returns the mock network name.
func (m *MockProvider) GetNetwork() string {
	return m.network
}

// GetRawTransaction returns the raw hex of a stored transaction.
func (m *MockProvider) GetRawTransaction(txid string) (string, error) {
	// Check raw transactions first (populated by broadcast)
	if raw, ok := m.rawTransactions[txid]; ok {
		return raw, nil
	}
	tx, ok := m.transactions[txid]
	if !ok {
		return "", fmt.Errorf("MockProvider: transaction %s not found", txid)
	}
	if tx.Raw == "" {
		return "", fmt.Errorf("MockProvider: transaction %s has no raw hex", txid)
	}
	return tx.Raw, nil
}

// GetFeeRate returns the configured fee rate (default 100 sat/KB).
func (m *MockProvider) GetFeeRate() (int64, error) {
	return m.feeRate, nil
}

// SetFeeRate sets the fee rate returned by GetFeeRate (for testing).
func (m *MockProvider) SetFeeRate(rate int64) {
	m.feeRate = rate
}

// ---------------------------------------------------------------------------
// Deterministic mock hash (produces a 64-char hex string like a txid)
// ---------------------------------------------------------------------------

func mockHash64(input string) string {
	h0 := uint32(0x6a09e667)
	h1 := uint32(0xbb67ae85)
	h2 := uint32(0x3c6ef372)
	h3 := uint32(0xa54ff53a)

	for i := 0; i < len(input); i++ {
		c := uint32(input[i])
		h0 = imul32(h0^c, 0x01000193)
		h1 = imul32(h1^c, 0x01000193)
		h2 = imul32(h2^c, 0x01000193)
		h3 = imul32(h3^c, 0x01000193)
	}

	parts := []uint32{h0, h1, h2, h3, h0 ^ h2, h1 ^ h3, h0 ^ h1, h2 ^ h3}
	var sb strings.Builder
	for _, p := range parts {
		fmt.Fprintf(&sb, "%08x", p)
	}
	return sb.String()
}

// imul32 multiplies two uint32 values, matching JavaScript's Math.imul semantics
// (32-bit wrapping multiplication).
func imul32(a, b uint32) uint32 {
	return a * b
}

File: packages/runar-zig/src/sdk_provider.zig
const std = @import("std");
const bsvz = @import("bsvz");
const types = @import("sdk_types.zig");

// ---------------------------------------------------------------------------
// Provider interface
// ---------------------------------------------------------------------------

/// Provider abstracts blockchain access for UTXO lookup and broadcast.
pub const Provider = struct {
    ptr: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        getTransaction: *const fn (ctx: *anyopaque, allocator: std.mem.Allocator, txid: []const u8) ProviderError!types.TransactionData,
        broadcast: *const fn (ctx: *anyopaque, allocator: std.mem.Allocator, tx_hex: []const u8) ProviderError![]u8,
        getUtxos: *const fn (ctx: *anyopaque, allocator: std.mem.Allocator, address: []const u8) ProviderError![]types.UTXO,
        getContractUtxo: *const fn (ctx: *anyopaque, allocator: std.mem.Allocator, script_hash: []const u8) ProviderError!?types.UTXO,
        getNetwork: *const fn (ctx: *anyopaque) []const u8,
        getFeeRate: *const fn (ctx: *anyopaque) ProviderError!i64,
        getRawTransaction: *const fn (ctx: *anyopaque, allocator: std.mem.Allocator, txid: []const u8) ProviderError![]u8,
    };

    pub fn getTransaction(self: Provider, allocator: std.mem.Allocator, txid: []const u8) ProviderError!types.TransactionData {
        return self.vtable.getTransaction(self.ptr, allocator, txid);
    }

    pub fn broadcast(self: Provider, allocator: std.mem.Allocator, tx_hex: []const u8) ProviderError![]u8 {
        return self.vtable.broadcast(self.ptr, allocator, tx_hex);
    }

    pub fn getUtxos(self: Provider, allocator: std.mem.Allocator, address: []const u8) ProviderError![]types.UTXO {
        return self.vtable.getUtxos(self.ptr, allocator, address);
    }

    pub fn getContractUtxo(self: Provider, allocator: std.mem.Allocator, script_hash: []const u8) ProviderError!?types.UTXO {
        return self.vtable.getContractUtxo(self.ptr, allocator, script_hash);
    }

    pub fn getNetwork(self: Provider) []const u8 {
        return self.vtable.getNetwork(self.ptr);
    }

    pub fn getFeeRate(self: Provider) ProviderError!i64 {
        return self.vtable.getFeeRate(self.ptr);
    }

    pub fn getRawTransaction(self: Provider, allocator: std.mem.Allocator, txid: []const u8) ProviderError![]u8 {
        return self.vtable.getRawTransaction(self.ptr, allocator, txid);
    }
};

pub const ProviderError = error{
    NotFound,
    BroadcastFailed,
    NetworkError,
    OutOfMemory,
};

// ---------------------------------------------------------------------------
// MockProvider — in-memory provider for testing
// ---------------------------------------------------------------------------

/// MockProvider is an in-memory provider for unit tests and local development.
pub const MockProvider = struct {
    allocator: std.mem.Allocator,
    utxos: std.StringHashMap(std.ArrayListUnmanaged(types.UTXO)),
    raw_transactions: std.StringHashMap([]const u8),
    broadcast_count: u32 = 0,
    broadcasted_txs: std.ArrayListUnmanaged([]const u8),
    network: []const u8,
    fee_rate: i64 = 100,

    pub fn init(allocator: std.mem.Allocator, network: []const u8) MockProvider {
        return .{
            .allocator = allocator,
            .utxos = std.StringHashMap(std.ArrayListUnmanaged(types.UTXO)).init(allocator),
            .raw_transactions = std.StringHashMap([]const u8).init(allocator),
            .broadcasted_txs = .empty,
            .network = if (network.len > 0) network else "testnet",
        };
    }

    pub fn deinit(self: *MockProvider) void {
        var utxo_it = self.utxos.iterator();
        while (utxo_it.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
            for (entry.value_ptr.items) |*u| u.deinit(self.allocator);
            entry.value_ptr.deinit(self.allocator);
        }
        self.utxos.deinit();

        var raw_it = self.raw_transactions.iterator();
        while (raw_it.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
            self.allocator.free(entry.value_ptr.*);
        }
        self.raw_transactions.deinit();

        for (self.broadcasted_txs.items) |tx| {
            self.allocator.free(tx);
        }
        self.broadcasted_txs.deinit(self.allocator);
    }

    /// Add a UTXO for the given address.
    pub fn addUtxo(self: *MockProvider, address: []const u8, utxo: types.UTXO) !void {
        const key = try self.allocator.dupe(u8, address);
        errdefer self.allocator.free(key);
        const gop = try self.utxos.getOrPut(key);
        if (gop.found_existing) {
            self.allocator.free(key);
        } else {
            gop.value_ptr.* = .empty;
        }
        try gop.value_ptr.append(self.allocator, try utxo.clone(self.allocator));
    }

    /// Add a raw transaction hex by txid.
    pub fn addRawTransaction(self: *MockProvider, txid: []const u8, raw_hex: []const u8) !void {
        const key = try self.allocator.dupe(u8, txid);
        errdefer self.allocator.free(key);
        const val = try self.allocator.dupe(u8, raw_hex);
        try self.raw_transactions.put(key, val);
    }

    /// Get the list of broadcasted transaction hex strings.
    pub fn getBroadcastedTxs(self: *const MockProvider) []const []const u8 {
        return self.broadcasted_txs.items;
    }

    /// Return a Provider interface backed by this MockProvider.
    pub fn provider(self: *MockProvider) Provider {
        return .{
            .ptr = @ptrCast(self),
            .vtable = &vtable,
        };
    }

    const vtable = Provider.VTable{
        .getTransaction = getTransactionImpl,
        .broadcast = broadcastImpl,
        .getUtxos = getUtxosImpl,
        .getContractUtxo = getContractUtxoImpl,
        .getNetwork = getNetworkImpl,
        .getFeeRate = getFeeRateImpl,
        .getRawTransaction = getRawTransactionImpl,
    };

    fn getTransactionImpl(_: *anyopaque, _: std.mem.Allocator, _: []const u8) ProviderError!types.TransactionData {
        return ProviderError.NotFound;
    }

    fn broadcastImpl(ctx: *anyopaque, allocator: std.mem.Allocator, tx_hex: []const u8) ProviderError![]u8 {
        const self: *MockProvider = @ptrCast(@alignCast(ctx));
        const stored_hex = allocator.dupe(u8, tx_hex) catch return ProviderError.OutOfMemory;
        self.broadcasted_txs.append(self.allocator, stored_hex) catch {
            allocator.free(stored_hex);
            return ProviderError.OutOfMemory;
        };
        self.broadcast_count += 1;

        // Generate deterministic fake txid
        const prefix = if (tx_hex.len > 16) tx_hex[0..16] else tx_hex;
        const fake_txid = mockHash64(allocator, self.broadcast_count, prefix) catch return ProviderError.OutOfMemory;

        // Store raw hex for later retrieval
        const txid_key = allocator.dupe(u8, fake_txid) catch return ProviderError.OutOfMemory;
        const raw_val = allocator.dupe(u8, tx_hex) catch {
            allocator.free(txid_key);
            return ProviderError.OutOfMemory;
        };
        self.raw_transactions.put(txid_key, raw_val) catch {
            allocator.free(txid_key);
            allocator.free(raw_val);
            return ProviderError.OutOfMemory;
        };

        return fake_txid;
    }

    fn getUtxosImpl(ctx: *anyopaque, allocator: std.mem.Allocator, address: []const u8) ProviderError![]types.UTXO {
        const self: *MockProvider = @ptrCast(@alignCast(ctx));
        const list = self.utxos.get(address) orelse return allocator.alloc(types.UTXO, 0) catch return ProviderError.OutOfMemory;
        var result = allocator.alloc(types.UTXO, list.items.len) catch return ProviderError.OutOfMemory;
        for (list.items, 0..) |u, i| {
            result[i] = u.clone(allocator) catch return ProviderError.OutOfMemory;
        }
        return result;
    }

    fn getContractUtxoImpl(_: *anyopaque, _: std.mem.Allocator, _: []const u8) ProviderError!?types.UTXO {
        return null;
    }

    fn getNetworkImpl(ctx: *anyopaque) []const u8 {
        const self: *MockProvider = @ptrCast(@alignCast(ctx));
        return self.network;
    }

    fn getFeeRateImpl(ctx: *anyopaque) ProviderError!i64 {
        const self: *MockProvider = @ptrCast(@alignCast(ctx));
        return self.fee_rate;
    }

    fn getRawTransactionImpl(ctx: *anyopaque, allocator: std.mem.Allocator, txid: []const u8) ProviderError![]u8 {
        const self: *MockProvider = @ptrCast(@alignCast(ctx));
        const raw = self.raw_transactions.get(txid) orelse return ProviderError.NotFound;
        return allocator.dupe(u8, raw) catch return ProviderError.OutOfMemory;
    }
};

// ---------------------------------------------------------------------------
// Deterministic mock hash (produces a 64-char hex string like a txid)
// ---------------------------------------------------------------------------

fn mockHash64(allocator: std.mem.Allocator, count: u32, prefix: []const u8) ![]u8 {
    var h0: u32 = 0x6a09e667;
    var h1: u32 = 0xbb67ae85;
    var h2: u32 = 0x3c6ef372;
    var h3: u32 = 0xa54ff53a;

    // Mix in count
    const count_str_buf = std.fmt.allocPrint(allocator, "mock-broadcast-{d}-", .{count}) catch return error.OutOfMemory;
    defer allocator.free(count_str_buf);
    for (count_str_buf) |c| {
        h0 = imul32(h0 ^ @as(u32, c), 0x01000193);
        h1 = imul32(h1 ^ @as(u32, c), 0x01000193);
        h2 = imul32(h2 ^ @as(u32, c), 0x01000193);
        h3 = imul32(h3 ^ @as(u32, c), 0x01000193);
    }
    for (prefix) |c| {
        h0 = imul32(h0 ^ @as(u32, c), 0x01000193);
        h1 = imul32(h1 ^ @as(u32, c), 0x01000193);
        h2 = imul32(h2 ^ @as(u32, c), 0x01000193);
        h3 = imul32(h3 ^ @as(u32, c), 0x01000193);
    }

    const parts = [8]u32{ h0, h1, h2, h3, h0 ^ h2, h1 ^ h3, h0 ^ h1, h2 ^ h3 };
    var result = try allocator.alloc(u8, 64);
    var pos: usize = 0;
    for (parts) |p| {
        _ = std.fmt.bufPrint(result[pos .. pos + 8], "{x:0>8}", .{p}) catch unreachable;
        pos += 8;
    }
    return result;
}

fn imul32(a: u32, b: u32) u32 {
    return a *% b;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "MockProvider returns UTXOs and broadcasts" {
    const allocator = std.testing.allocator;
    var mock = MockProvider.init(allocator, "testnet");
    defer mock.deinit();

    try mock.addUtxo("addr1", .{
        .txid = "aabb",
        .output_index = 0,
        .satoshis = 1000,
        .script = "76a914",
    });

    var prov = mock.provider();

    const utxos = try prov.getUtxos(allocator, "addr1");
    defer {
        for (utxos) |*u| {
            var mu = u.*;
            mu.deinit(allocator);
        }
        allocator.free(utxos);
    }
    try std.testing.expectEqual(@as(usize, 1), utxos.len);
    try std.testing.expectEqual(@as(i64, 1000), utxos[0].satoshis);

    const txid = try prov.broadcast(allocator, "0100000000");
    defer allocator.free(txid);
    try std.testing.expectEqual(@as(usize, 64), txid.len);

    const fee_rate = try prov.getFeeRate();
    try std.testing.expectEqual(@as(i64, 100), fee_rate);
}
